### ISSUE TITLE
Production dependencies in a local development environment

### ISSUE THREAD

**Issue by iaindillingham** [opensafely-actions/python-action-template#30]
TL;DR: *How should we make the production dependencies available in a local development environment?*

At present, we do this as follows:

https://github.com/opensafely-actions/python-action-template/blob/618ef6a02e65a91684545c64ad1d1f7296f87859/requirements.dev.in#L4-L6

Notice that we read `python-docker`'s *requirements.in* rather than its *requirements.txt* because [`pip-compile`](https://github.com/jazzband/pip-tools/#readme) can't resolve [click](https://click.palletsprojects.com/en/8.0.x/). (There's nothing special about click, here; it could just as easily be another production dependency.) You can read more about this decision in #24. Crucially, I asserted:

> Whilst this means the local development environment doesn't contain the global production dependencies *exactly*, it nevertheless contains them *approximately*. My assertion is that "approximately" is "good enough" for most users of this template repo.

Uh-oh; it turns out that it isn't 🤦🏻‍♂️. Having Plotly v5.2.1 in development and Plotly v4.5.0 in production is the difference between rendering charts and not rendering charts. You can read more about this in opensafely-actions/cohort-report#20.

So, we shouldn't read `python-docker`'s *requirements.in* and we can't read its *requirements.txt*. We could ask a developer of a Python reusable action to pin production dependencies, ideally in *requirements.prod.in*; they could then run `pip-compile ...` to create *requirements.prod.txt*, which could then be required (`-r`) by *requirements.dev.in*. However, pinning in *requirements.prod.in* isn't the `pip-compile` way; and requires the developer to manage production dependencies. My guess is that the developer, who is likely to come from a research background, will be unsure of why they are required to do so.

---

For the curious, some context...

Python reusable actions are executed within the Python *runtime*, or the image that's created by [`python-docker`](https://github.com/opensafely-core/python-docker). This means that Python reusable actions are executed by the version of Python available within the Python runtime, which is currently Python 3.8.2 (how do we know this? See below!). This also means that Python reusable actions have access to the dependencies given by [*requirements.txt*](https://github.com/opensafely-core/python-docker/blob/main/requirements.txt), which are derived from [*requirements.in*](https://github.com/opensafely-core/python-docker/blob/main/requirements.in). (As well as listing the dependencies, the former also pins them to their exact versions, e.g. `pandas==1.0.1`.)

The dependencies given by *requirements.txt* are a Python reusable action's *production* dependencies. A Python reusable action can call none, some, or all of these dependencies when running in production (i.e. when executed within the Python runtime). Importantly, a Python reusable action cannot call a package that is not either a direct dependency or a transitive dependency (a dependency of a dependency; it's 🐢 🐢  all the way down).

When developing a Python reusable action, it's useful to have a local development environment running Python 3.8 with the development dependencies installed. (The development dependencies are given by [*requirements.dev.txt*](https://github.com/opensafely-actions/python-action-template/blob/main/requirements.dev.txt), which are derived from [*requirements.dev.in*](https://github.com/opensafely-actions/python-action-template/blob/main/requirements.dev.in).) This is because the development dependencies help improve code quality. Indeed, it's almost essential to have a local development environment because the development dependencies are not available in production; whilst they are available through the workflow GH action, a development cycle that involved writing code locally, pushing to GH, and then waiting for feedback from the workflow GH action would be inefficient.

*How do we know which version of Python is available within the Python runtime?*

`python-docker` references an image called `base-action`, which is created by [`base-docker`](https://github.com/opensafely-core/base-docker). This image references an image called `ubuntu:20.04`; unsurprisingly, this image is of Ubuntu 20.04, or *focal*, which is a LTS (Long-Term Support) release. `python-docker` creates a virtual environment to run a Python reusable action; it does this by calling `python3`. [The default `python3` on focal is 3.8.2](https://packages.ubuntu.com/focal/python3).

**Comment by inglesp** [comment]
Can we drop pip-compile completely?  That is, can we say that a reusable action must use the requirements.txt of the base image?  (And enforce this somehow...)

But this only postpones the problem -- what happens to actions when we upgrade the base image?

**Comment by iaindillingham** [comment]
We still need `pip-compile` to compile the development dependencies.* If the development dependencies referenced the production dependencies (`-r`), then I think `pip-compile` would still complain that it was unable to resolve click. If they didn't, and if we installed production over development or vice versa, then we might find hard-to-debug conflicts.

*It's worth questioning this assertion; maybe we don't need to compile the development dependencies?
